File: src/Billing/Mediation/Rating/RateCaching/DateRange.cs
// Ported VERBATIM from legacy LibraryExtensions.DateRange (OtherExtensions.cs) — the day key of the
// RateCache's DateRangeWiseRateDic. Value-equality on (StartDate, EndDate) + the EqualityComparer the
// dictionary uses.
#nullable disable
using System;
using System.Collections.Generic;

namespace LibraryExtensions
{
    public class DateRange : IEquatable<DateRange>
    {
        public DateTime StartDate = new DateTime(1, 1, 1);
        public DateTime EndDate = new DateTime(1, 1, 1);

        public DateRange() { }

        public DateRange(DateTime startDateTime, DateTime endDateTime)
        {
            this.StartDate = startDateTime;
            this.EndDate = endDateTime;
        }

        public override string ToString() =>
            this.StartDate.ToString("yyyy-MM-dd HH:mm:ss") + " to " + this.EndDate.ToString("yyyy-MM-dd HH:mm:ss");

        public List<DateTime> GetInvolvedHours()
        {
            var hours = new List<DateTime>();
            var current = new DateTime(StartDate.Year, StartDate.Month, StartDate.Day, StartDate.Hour, 0, 0);
            while (current <= EndDate)
            {
                hours.Add(current);
                current = current.AddHours(1);
            }
            return hours;
        }

        public bool WithinRange(DateTime dateTime) => dateTime >= this.StartDate && dateTime < this.EndDate;

        public DateRange NewRangeByAddingDays(int daysToAdd) =>
            new() { StartDate = this.StartDate.AddDays(daysToAdd), EndDate = this.EndDate.AddDays(daysToAdd) };

        public override int GetHashCode()
        {
            unchecked
            {
                int hash = 17;
                hash = hash * 29 + this.StartDate.GetHashCode();
                hash = hash * 29 + this.EndDate.GetHashCode();
                return hash;
            }
        }

        public override bool Equals(object obj) => Equals(obj as DateRange);

        public bool Equals(DateRange obj) =>
            obj != null && obj.StartDate == this.StartDate && obj.EndDate == this.EndDate;

        public class EqualityComparer : IEqualityComparer<DateRange>
        {
            public bool Equals(DateRange x, DateRange y) =>
                x != null && y != null && x.StartDate == y.StartDate && x.EndDate == y.EndDate;

            public int GetHashCode(DateRange x)
            {
                unchecked
                {
                    int hash = 17;
                    hash = hash * 29 + x.StartDate.GetHashCode();
                    hash = hash * 29 + x.EndDate.GetHashCode();
                    return hash;
                }
            }
        }
    }
}

File: java/src/main/java/com/telcobright/billing/mediation/rating/ratecaching/DateRange.java
// Ported VERBATIM from legacy LibraryExtensions.DateRange (OtherExtensions.cs) — the day key of the
// RateCache's DateRangeWiseRateDic. Value-equality on (StartDate, EndDate) + the EqualityComparer the
// dictionary uses.
package com.telcobright.billing.mediation.rating.ratecaching;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class DateRange {
    public LocalDateTime StartDate = LocalDateTime.of(1, 1, 1, 0, 0);
    public LocalDateTime EndDate = LocalDateTime.of(1, 1, 1, 0, 0);

    private static final DateTimeFormatter Fmt = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    public DateRange() { }

    public DateRange(LocalDateTime startDateTime, LocalDateTime endDateTime) {
        this.StartDate = startDateTime;
        this.EndDate = endDateTime;
    }

    @Override
    public String toString() {
        return this.StartDate.format(Fmt) + " to " + this.EndDate.format(Fmt);
    }

    public List<LocalDateTime> GetInvolvedHours() {
        var hours = new ArrayList<LocalDateTime>();
        var current = LocalDateTime.of(StartDate.getYear(), StartDate.getMonthValue(), StartDate.getDayOfMonth(), StartDate.getHour(), 0, 0);
        while (!current.isAfter(EndDate)) {
            hours.add(current);
            current = current.plusHours(1);
        }
        return hours;
    }

    public boolean WithinRange(LocalDateTime dateTime) {
        return !dateTime.isBefore(this.StartDate) && dateTime.isBefore(this.EndDate);
    }

    public DateRange NewRangeByAddingDays(int daysToAdd) {
        DateRange r = new DateRange();
        r.StartDate = this.StartDate.plusDays(daysToAdd);
        r.EndDate = this.EndDate.plusDays(daysToAdd);
        return r;
    }

    @Override
    public int hashCode() {
        int hash = 17;
        hash = hash * 29 + this.StartDate.hashCode();
        hash = hash * 29 + this.EndDate.hashCode();
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        return equals(obj instanceof DateRange ? (DateRange) obj : null);
    }

    public boolean equals(DateRange obj) {
        return obj != null && Objects.equals(obj.StartDate, this.StartDate) && Objects.equals(obj.EndDate, this.EndDate);
    }

    // Mirrors the legacy IEqualityComparer<DateRange>. Retained for fidelity; Java's HashMap uses the
    // key's own equals()/hashCode() (identical logic), so this is not wired into the cache map directly.
    public static class EqualityComparer {
        public boolean equals(DateRange x, DateRange y) {
            return x != null && y != null && Objects.equals(x.StartDate, y.StartDate) && Objects.equals(x.EndDate, y.EndDate);
        }

        public int getHashCode(DateRange x) {
            int hash = 17;
            hash = hash * 29 + x.StartDate.hashCode();
            hash = hash * 29 + x.EndDate.hashCode();
            return hash;
        }
    }
}
